### DISCUSSION TITLE
GSoc 2022 projects

### DISCUSSION THREAD

**Discussion by pepeiborra** [haskell/haskell-language-server#2700] [Ideas]
@jaspervdj has reached out to encourage us to submit project ideas to https://github.com/haskell-org/summer-of-haskell 

Let's gather some proposals and add them as soon as possible.

Note that there's been some big changes this year:

1. There is now the choice of doing large projects as well as medium
projects (more or less full time vs part-time)
2. We are allowed to accept any contributors (no longer limited to students)

**Comment by pepeiborra** [comment]
## GHC 9.4 support

As agreed in https://github.com/haskell/haskell-language-server/discussions/2588 a top priority is to get HLS working with GHC 9.4, which should hopefully have a pre-rc out before the summer starts. 

@mpickering @wz1000 would either of you be able to mentor a student on this project?

**Reply by wz1000** [reply]
I'd be willing to mentor such a project.

**Reply by wz1000** [reply]
I'm working on this myself, so perhaps this is not the best project for a student.

**Comment by pepeiborra** [comment]
## Typed diagnostics 

GHC 9.4 comes with new support for typed diagnostics (and apparently, most of it is already available in 9.2). See also #2014 

@adinapoli @goldfirere would you be willing to mentor a student on this project?

**Reply by adinapoli** [reply]
Hello @pepeiborra , sorry for the delay but I have recently become dad of two and time is pretty limited these days 😉 I'll get back to you on this, hopefully this week.

**Reply by pepeiborra** [reply]
@adinapoli mentors do little or no work. I speak from experience, mentored 2 SoCs last year.

**Reply by adinapoli** [reply]
@pepeiborra right, but I am working part time these days as the babies tends to absorb most of my day. I am happy to act as a "counselor" and reply to questions or emails here and there, I am just worried that if I have to actually mentor somebody I might not be in a position of actually mentoring, making the whole experience rather poor 😿

**Reply by pepeiborra** [reply]
@drsooch would you be interested in co-mentoring with @adinapoli?

**Reply by drsooch** [reply]
Sorry I never saw this! Looks like this one wasn't picked up? But if the offer is still out there I'd be happy to help

**Comment by pepeiborra** [comment]
GSoC 2021 ideas: https://github.com/haskell/haskell-language-server/discussions/1237

**Comment by googleson78** [comment]
Sorry for butting in without being a core contributor or anything of the sort! And this is in no way any kind of demand, nor am I here to whine - it's honestly just what I feel is most important for the project (for myself, of course)!

The biggest issue for me with hls is how often it arbitrarily crashes/doesn't load. I think a lot of these cases might be due to stack, so I guess this might require some work (with someone) there. I'm sure there are other issues around this though - I've seen it just not work on a bare file for a reason I couldn't assess at the time, while immediately trying most of the debug things I knew. (fresh install in vscode)

In any case, what I want to express by rambling is that if it was up to me and I had the time/resources to invest, this would be my first priority - for me the HLS experience is already **superb** *when* it hasn't crashed or I'm waiting for it to maybe load something an indefinite amount of time.

**Reply by pepeiborra** [reply]
We need something more concrete than "Make HLS not crash" to ensure a successful project though. Would you or anyone else be willing to define that up and mentor (or apply as a student)?

**Reply by michaelpj** [reply]
Could I suggest a sub-problem? "Make HLS give very clear diagnostics about why it crashed, or the events leading up to a crash". AKA "make our logging really good, including dark corners like cradle loading".

A big issue with the crashes is that they're *mysterious* crashes. If you looked in the log and it said "oh no, I can't handle XYZ, did you mean to do PQR?" that would make a huge difference.

**Reply by pepeiborra** [reply]
Agreed. Rather than logging, HLS should surface these problems via LSP diagnostics or LSP message dialogs. The log is often hard to retrieve, and the old advice to repro using the cli is not reliable.

**Reply by googleson78** [reply]
It's also not clear (to me) how to repro with the cli sometimes, as most crashes/hangs happen when I make a change to a file and it tries to reload it, but if I restart the whole thing it works fine.

**Reply by michaelpj** [reply]
I'm hoping to get to the point soon-ish where *all* error-level logs get surfaced to the user in that way. So hopefully we can solve those "more logs" and "more accessible logs" problems orthogonally.

**Reply by jaspervdj** [reply]
Anyone willing to write this up as an idea?

**Reply by michaelpj** [reply]
@eddiemundo ? Would you be interested in mentoring someone working on this?

**Reply by eddiemundo** [reply]
I could mentor somebody but I kind of think that the logging refactoring is a small project in the sense that in my mind it would take like 10 hours max although my estimations aren't great. The fuzzy part is figuring out what messages need to be made more accessible, how to make them accessible, and where to put additional logging (to maybe even catch possibly silent crashes) although I have some leads on the last one.

**Reply by michaelpj** [reply]
So sounds like a small project at most. But I think there could be a lot to do if we got into the fuzzy stuff. Anyway, up to you, or anyone else who's keen to be a mentor!

**Comment by jaspervdj** [comment]
Thanks for taking charge of this @pepeiborra!

**Comment by michaelpj** [comment]
There are some LSP protocol features that we don't use yet. Implementing some of those could be a project (I'd be happy to mentor).

- Change annotations. These let you annotate pieces of edits with notes explaining what they do, and in some clients may let the user review them before accepting. For example, I think this would let us annotate each hlint fix in the "apply all fixes" edit with the hint that it is addressing, making it more obvious what it has done.
- Semantic highlighting.
- Folding ranges.
- Linked editing. This is actually quite nice: it's a lightweight way of doing document-local renamings for e.g. local variables.
- Completion/code action/code lens resolving. This is definitely the most complicated, since it potentially requires state tracking on the part of the server. But if we could do it well then it might lead to improved performance. Completions would probably be highest priority.
- Document links. These would be very nice if we could do them for e.g. type references in Haddock.

Writing this list has made me realise we should have some issues for these regardless, I'll make some.

**Reply by jaspervdj** [reply]
@michaelpj This sounds really nice, especially because students could pick which features they'd implement based on their experience, and we could potentially even have multiple students and split it up if we get enough project slots.  Are you okay with me adding this as an idea with you as the contact person / mentor?

**Reply by michaelpj** [reply]
Sure. I'm not sure I'd have the bandwidth to mentor more than one, but there's probably more than a single project in there, so if someone else wants to help we could maybe take more than one!

**Reply by jaspervdj** [reply]
Thanks, I've added it here: https://summer.haskell.org/ideas.html#hls-lsp-features -- feel free to send me any corrections or update.  And of course, there's no obligation to take on more than one student at all.

**Comment by pepeiborra** [comment]
Adding support for dot notation in GHC 9.2 (see #2732) would make for a small SoC project. @michaelpj would you be willing to mentor it?

**Reply by michaelpj** [reply]
Potentially, although I don't know much about the HLS completions code. I'm also not sure if I should sign up to mentor multiple projects - I haven't mentored one before and I don't want to overcommit myself.

**Reply by pepeiborra** [reply]
I'm happy to mentor this one if your other project gets a student.

**Reply by michaelpj** [reply]
Works for me!

**Reply by pepeiborra** [reply]
@michaelpj would you be able to submit the project to https://summer.haskell.org/ideas.html ?

**Reply by michaelpj** [reply]
https://github.com/haskell-org/summer-of-haskell/pull/153

**Reply by coltenwebb** [reply]
I'm a student interested in doing this. Would either of you mind looking at a draft of my proposal if I send it via email?

**Reply by michaelpj** [reply]
Sure!

**Comment by bradrn** [comment]
For a while now I have been discussing the addition of GHCi integration into HLS (see https://github.com/haskell/haskell-language-server/issues/477). It occurs to me that this might make a great shorter SoC project, but it would need a mentor.

**Reply by michaelpj** [reply]
I think this would be cool, but I think it's probably a lot of work, I'd expect it to be a longer project at least...

**Reply by michaelpj** [reply]
And it's probably a bit late to get into the program now.

**Reply by bradrn** [reply]
> I think this would be cool, but I think it's probably a lot of work, I'd expect it to be a longer project at least...

Well, much of the work has already been done by @phoityne (see the comments of the linked issue). I suspect it should be possible to at least get a working demo in 4–5 weeks (which is the amount of time I would have, being in the Southern Hemisphere).

> And it's probably a bit late to get into the program now.

I’m pretty sure there’s a few more days yet before applications close…

**Reply by michaelpj** [reply]
Oh I see, I didn't realise you were interested yourself, I thought we'd also have to find students. We'd still need a mentor, though, and I think I'm booked, anyone else interested?

**Reply by bradrn** [reply]
Alas, upon further investigation it appears I will be unable to participate in GSoC this year due to time constraints. I wouldn’t mind if another student wants to do this project.
